src/team_lib/mhba/xccl_mhba_team.c

```diff
@@ -95,8 +95,9 @@ static ucs_status_t rcache_reg_mr(void *context, ucs_rcache_t *rcache,void *arg,
     size_t length             = (size_t)(rregion->super.end - rregion->super.start);
     xccl_mhba_reg_t* mhba_reg = xccl_rcache_ucs_get_reg_data(rregion);
     mhba_reg->region = rregion;
-    int* mem_flags = (int*) arg;
-    mhba_reg->mr = ibv_reg_mr(team->node.shared_pd, addr, length, *mem_flags);
+    int* change_flag = (int*) arg;
+    *change_flag = 1;
```

Comment:
So, you decide if you need to update the key only based on the rcache registration callback execution? I don't get it: how it works? If the buffer was registered and not released by user - it is still in rcache. But the mkey is not longer relevant. Here is an example:
1. user calls alltoall with buffer = BUF1 (assume msgsize = 4 for simplicity)
2. BUF1 will get into UCS rcache. Mkey is created for BUF1/4b
3. Now, user calls alltoall with BUF2 (assume same 4b)
4. BUF2 gets into UCS rcache as well. Mkey is now created for BUF2/4b
5. Not user may call alltoall with BUF1 again, it will be found in UCS rcache and hence registration callback is not called, you don't set the change_flag. Thus, you will use wrong Mkey from BUF2/4b.

Second question: how will you handle the case when BUF1 is called second time but with different msgsize (say, 2bytes). It will be found in rcache but Mkey is wrong.

Need a better logic. You need to write some kind of function that builds a "key" based on buffer/msgsize and then check if the "key" from previous alltoall call is exactly same as next one. Only in this case you can re-use mkey, otherwise need an update.

Reply:
changed